content/concepts/peer-routing.md: add first draft (#147)
libp2p/docs: content/concepts/peer-routing.md

weight: 5
---

> Libp2p primarily uses Kademlia DHT for peer discovery. Let's dive a little bit into how Kademlia works.

Reviewer: It is one mechanism among many. E.g. ethereum 2 is not using it at all.

Author: fixed in new commit, but still not sure if it sounds good, since I don't know yet many other ways to discover peers 🙃

## Static DHT

To grow intuition about the Kademlia let's assume for now that our DHT is static. No one joins DHT and no one leaves.

### System Descritpion

Each peer has their own peer ID. Kademlia provides an efficient lookup algorithm that locates successively “closer” nodes to any desired ID.

> In Kademlia peers are key-value pairs in the DHT. On practice peer ID stored as a key and [address](/concepts/addressing/) (or IP address) of a peer stored as a value.

Reviewer: ```suggestion
In Kademlia peers are key-value pairs in the DHT. Conceptually peer ID stored as a key and [address](/concepts/addressing/) (or IP address) of a peer stored as a value.
```

Author: 👍

> Kademlia treats peers as leaves in a binary tree. Peer IDs in leaves represented as binary numbers. Notice, that we do not need to store the whole peer id to specify the peer. We can storeƒ only their prefix if prefix is unique.

Reviewer: We actually hash the IDs with sha256 to get a uniform key space. That said, I am not sure it needs mentioning here.

Author: I think, that it is good to mention this and then show hash is being stored as binary

That is because I also think that for the content-routing it would be good to explain distances

![Kademlia-system](Kademlia-system.png)

Red node here has prefix `001`, go from the top to the bottom by edges. Let's say it' Bob. Alice here has prefix `11101` - she is blue node. Yellow leaves are other peers. For any given node, Kademlia divides the whole binary tree into a series of successively lower subtrees that don’t contain the given node. The highest subtree consists of the half of the binary tree not containing the node. The next subtree consists of the half of the remaining tree not containing the node, and so forth. With example it'll be more clear. You can see here for node `001` subtrees are circled and consist of all nodes with prefixes `1`, `01`, `000`.

The Kademlia protocol ensures that every node knows of at least one node in each of its subtrees, if that subtree contains a node. This guarantees that any node can be located by it's peer ID.

Let's see how Bob can locate Alice in this example. Bob knows Alices peerID, but he needs to get the value - Alices IP address. From subtree with prefix `1` Bob knows peer with prefix `100`. He sends a message to their address, where he asks for Alices address.

> ![1-hop](1-hop.png)

Reviewer: Mhh, these don't seem to render in GitHubs markdown view. Do they do so on your end?

Author: No, they don't for me
I was doing the same thing as in [pubsub](https://github.com/libp2p/docs/blob/master/content/concepts/publish-subscribe.md) article, so it shows up [here](https://bafybeidklhiwzma7r3vufxkvoym3pq6pv2l2hnmnu5uynvpqfjnbqmd2bi.on.fleek.co/concepts/peer-routing/)

![3-hop](3-hop.png)

Peer `1111` stores Alices peer id and can send the key to the Bob! This is how routing works and here is full scheme:

![final-hop](final-hop.png)

Peer `1111` sends response directly to Bob, since each request in this chain had Bob's address.

### Routing table

The place where one node stores info about other nodes called routing table. The routing table is a binary tree whose leaves are `k-buckets`. Let's boil it down.

What are `k-buckets`? Remember our initial tree? Here it is. ![Kademlia-system](Kademlia-system.png) See this circled sub-trees? Also, remember that Kademlia protocol ensures that every node knows of at least one node in each of its subtrees, if that subtree contains a node. `k-bucket` of a sub-tree is a set of that nodes. Once again: from each subtree we take only those nodes that we know - and we call it `k-bucket`. In out case `k-buckets` for Bob look like this

![k-buckets](k-buckets.png)

Each `k-bucket` covers some range of the ID space, and together the `k-buckets` cover the entire peer ID space with no overlap.

## Dynamic DHT

> Okay, that was it for static DHT. And the gap from our current state to a fully dynamic DHT is not as far away as it seems. Let us first consider how we can extend this protocol to support computers sporadically leaving.

Reviewer: Cool to only introduce this complexity later on in the article :+1:

Author: Oh I also like this, it wasn't originally my idea though

I borrowed it from [here](https://codethechange.stanford.edu/guides/guide_kademlia.html)
(and I referred to this source at the end :))

### Leaving

Let us assume that we have some known parameter `k` that represents the number of computers such that it is extremely unlikely that all of those computers will leave the network in the same hour. For example, the authors of the original Kademlia paper experimentally determined that their `k` should be 20.

Next, our `k-bucket` should, instead of storing just 1 computer, store `k` computers within that `k-bucket’s` range. Thus, with high likelihood, there will always be at least one computer online in each of a computer’s `k-buckets`.

And, when hopping, instead of asking one computer from `k-bucket` we will be asking `k`. Some may not reply, but not all of them.

> To check if node is still alive we can ping it periodically. If node doesn't respond in the allotted time it will be evicted from our `k-bucket`.

Reviewer: Not quite. It will be marked as non-responding and eventually replaced once we have a better node.

Author: fixed

How do we become aware of new computers that could fit into our unfilled `k-buckets`? We could perform lookups on random ids within a `k-bucket’s` id range (the range is defined by all leaves that are direct descendants of the `k-bucket’s` corresponding internal node, which is not leaf, in the complete tree) and thus learn about computers within that `k-bucket` if they exist. This is called a _bucket refresh_.

### Joining

To join the DHT new computer `Eve` must know at least on IP address of computer which is in the network. Then `Eve` will perform a lookup of herself to get closest known computers which are already aware of `Eve`. And after that `Eve` will perform a _bucket refresh_ for all `k-buckets` from the closest known computers to `Eve`. This will populate the `Eve's` routing table based on the correctness of the lookup algorithm and the fact that the routing tables of other computers are sufficiently populated. In addition, all computers that are being queried by `Eve` will become aware of `Eve` in the process and thus will have the opportunity to insert `Eve` into their own `k-buckets` if their corresponding bucket is not already full.

> ## More information

Reviewer: First thing I would link to is the Kademlia paper. Second would be the libp2p Kademlia specification.

- https://github.com/libp2p/specs/blob/master/kad-dht/README.md
- https://pdos.csail.mit.edu/~petar/papers/maymounkov-kademlia-lncs.pdf

Author: And added that

Author: I should have guessed to add the specs :))

weight: 5
---

One of the many mechanisms that libp2p uses for peer routing is Kademlia DHT. We will dive into how it works in this article, this will help you further understand more advanced topics.

## Static DHT

> To grow intuition about the Kademlia let's assume for now that our DHT is static. No one joins DHT and no one leaves.

Reviewer 2: ```suggestion
In this example, we are going to assume for now that our DHT is static, in hat one joins DHT and no one leaves.
```

### System Descritpion

Each peer has their own peer ID. Kademlia provides an efficient lookup algorithm that locates successively “closer” nodes to any desired ID.

In Kademlia peers are key-value pairs in the DHT. Conceptually peer ID stored as a key and [address](/concepts/addressing/) (or IP address) of a peer stored as a value.

Kademlia treats peers as leaves in a binary tree. We store hashes of Peer IDs in leaves, which are represented as binary numbers. Notice, that we do not need to store the whole hash to specify the peer. We can store only prefix if prefix is unique.

Leaves on the tree that are not peers represent [content](/concepts/content-routing/) which is stored by some close nodes.

![Kademlia-system](Kademlia-system.png)

> Red node here has prefix `001`, go from the top to the bottom by edges. Let's say it' Bob. Alice here has prefix `11101` - she is blue node. Yellow leaves are other peers. For any given node, Kademlia divides the whole binary tree into a series of successively lower subtrees that don’t contain the given node. The highest subtree consists of the half of the binary tree not containing the node. The next subtree consists of the half of the remaining tree not containing the node, and so forth. With example it'll be more clear. You can see here for node `001` subtrees are circled and consist of all nodes with prefixes `1`, `01`, `000`.

Reviewer 2: ```suggestion
The red node in this diagram has the prefix `001`. Traversing this graph from the top to the bottom along the edges, you will notice a blue node with the prefix `11101`, as well as yellow leaves which are other peers. 

For any given node, Kademlia divides the whole binary tree into a series of successively lower subtrees. The highest subtree consists of the half of the binary tree not containing the original node. The next subtree consists of the half of the remaining tree not containing the next highest node, and so on. 

You can see here for node `001`  there are three levels of subtrees are also circle,  consisting of all nodes with prefixes `1` (highest level subtree) , `01`, `000` (lowest level subtree).
```

The Kademlia protocol ensures that every node knows of at least one node in each of its subtrees, if that subtree contains a node. This guarantees that any node can be located by it's peer ID.

> Let's see how Bob can locate Alice in this example. Bob knows Alices peerID, but he needs to get the value - Alices IP address. From subtree with prefix `1` Bob knows peer with prefix `100`. He sends a message to their address, where he asks for Alices address.

Reviewer 2: ```suggestion
Let's see how once can locate in an example. We will call the red node 'Bob', with the prefix  `001.` We will call the yellow node Alice, with the prefix `01`

Bob can locate Alice in this example.  Bob knows Alice's peerID, but he needs to get Alice's location address. From the subtree with the prefix `1` Bob knows a peer with prefix `101`. He sends a message to their address, where he asks for Alice's address.
```

Reviewer 2: I'm not entirely sure what you mean by this. The words address, key, location address keep being used interchangeably. I recommend you review all of this paragraph to accurately describe what is being located. Is it a PeerID? IP Address?

Author: Yeah, i'm sorry for this mess.

Here DHT is a storage for key-value pairs. 
In our case key is Alice's peer ID. Bob knows it, and he needs to get the value.
Value here is Alice's IP address, so then Bob will be able to communicate directly with this IP address (and this is Alice)

I added "a brief reminder" paragraph in my current commit and I hope it'll make it clearer.

> ![1-hop](1-hop.png)

Reviewer 2: @poolsar42 I think you have to upload the asset into the correct folder, like these other images: https://github.com/libp2p/docs/tree/master/content/concepts/publish-subscribe

Author: @walkerlj0 I think I did the same thing: you can see that for publish-subscribe concept "publish-subscribe" was created, and I just created "peer-routing" folder for content for this article.
Please hmu if you feel that I'm doing it wrong, because [here](https://bafybeidklhiwzma7r3vufxkvoym3pq6pv2l2hnmnu5uynvpqfjnbqmd2bi.on.fleek.co/concepts/peer-routing/) it's rendering just fine (this is check for fleek build)

Reviewer 2: I was able to fetch this branch and render the images, (they look great and add so much to the article!) @poolsar42

> Peer `100` doesn't have Alice's key. But they have key of node `110` from their subtree which locates Alice better! And we are doing second hop:

Reviewer 2: ```suggestion
Peer `100` doesn't have Alice's address, so they reroute and ask the peer located at the node with the prefix `110` from their subtree.
```

![2-hop](2-hop.png)

> Peer `110` also doesn't have Alice's key, but they again locate Alice better and send request to node with prefix `1111`.

Reviewer 2: ```suggestion
In this example, peer `110` also doesn't have Alice's address, so they send request to another node, with the prefix `1111`.
```

![3-hop](3-hop.png)

Peer `1111` stores Alices peer id and can send the key to the Bob! This is how routing works and here is full scheme:

![final-hop](final-hop.png)

Peer `1111` sends response directly to Bob, since each request in this chain had Bob's address.

### Routing table

> The place where one node stores info about other nodes called routing table. The routing table is a binary tree whose leaves are `k-buckets`. Let's boil it down.

Reviewer 2: ```suggestion
The place where one node stores info about other nodes is called a routing table. The routing table is a binary tree whose leaves are `k-buckets`. 
```

> What are `k-buckets`? Remember our initial tree? Here it is. ![Kademlia-system](Kademlia-system.png) See this circled sub-trees? Also, remember that Kademlia protocol ensures that every node knows of at least one node in each of its subtrees, if that subtree contains a node. `k-bucket` of a sub-tree is a set of that nodes. Once again: from each subtree we take only those nodes that we know - and we call it `k-bucket`. In out case `k-buckets` for Bob look like this

Reviewer 2: ```suggestion
### k-buckets
Take a look at the initial tree:
 ![Kademlia-system](Kademlia-system.png) 
See this circled sub-trees? Remember that the Kademlia protocol ensures that every node knows of (and can locate) at least one node in each of its subtrees. If that subtree contains a node. `k-bucket` of a sub-tree is a set of that nodes. Once again: from each subtree we take only those nodes that we know - and we call it `k-bucket`. In out case `k-buckets` for Bob look like this
```

![k-buckets](k-buckets.png)

Each `k-bucket` covers some range of the ID space, and together the `k-buckets` cover the entire peer ID space with no overlap.

## Dynamic DHT

Okay, that was it for static DHT. And the gap from our current state to a fully dynamic DHT is not as far away as it seems. Let us first consider how we can extend this protocol to support computers sporadically leaving.

> ### Leaving

Reviewer 2: ```suggestion
### Peers Leaving 
```

> Let us assume that we have some known parameter `k` that represents the number of computers such that it is extremely unlikely that all of those computers will leave the network in the same hour. For example, the authors of the original Kademlia paper experimentally determined that their `k` should be 20.

Reviewer 2: ```suggestion
Let us assume that we have some known parameter `k` that represents some given number of computers.  This number should be a value such that it is extremely unlikely that all of those computers will leave the network in the same hour. The original Kademlia paper that the `k` value should be 20.
```

Next, our `k-bucket` should, instead of storing just 1 computer, store `k` computers within that `k-bucket’s` range. Thus, with high likelihood, there will always be at least one computer online in each of a computer’s `k-buckets`.

> And, when hopping, instead of asking one computer from `k-bucket` we will be asking `k`. Some may not reply, but not all of them.

Reviewer 2: ```suggestion
And, when hopping, instead of asking one computer from the `k-bucket` we will be asking `k` number of peers. Some may not reply, but not all.
```

> To check if node is still alive we can ping it periodically. If node doesn't respond in the allotted time it will be marked as non-responding and eventually replaced once we have a better one.

Reviewer 2: ```suggestion
To check if each node is still alive we can ping it periodically. If node doesn't respond in the allotted time it will be marked as non-responding and eventually replace it with another.
```

> How do we become aware of new computers that could fit into our unfilled `k-buckets`? We could perform lookups on random ids within a `k-bucket’s` id range (the range is defined by all leaves that are direct descendants of the `k-bucket’s` corresponding internal node, which is not leaf, in the complete tree) and thus learn about computers within that `k-bucket` if they exist. This is called a _bucket refresh_.

Reviewer 2: ```suggestion
How do we become aware of new computers that could fit into our unfilled `k-buckets`? We can perform lookups on random ids within a `k-bucket’s` id range (the range is defined by all leaves that are direct descendants of the `k-bucket’s` corresponding internal node, which is not a leaf, in the complete tree) and thus learn about other computers within that `k-bucket`, if they exist. This is called a _bucket refresh_.
```

> ### Joining

Reviewer 2: ```suggestion
### Peers Joining
```

> To join the DHT new computer `Eve` must know at least on IP address of computer which is in the network. Then `Eve` will perform a lookup of herself to get closest known computers which are already aware of `Eve`. And after that `Eve` will perform a _bucket refresh_ for all `k-buckets` from the closest known computers to `Eve`. This will populate the `Eve's` routing table based on the correctness of the lookup algorithm and the fact that the routing tables of other computers are sufficiently populated. In addition, all computers that are being queried by `Eve` will become aware of `Eve` in the process and thus will have the opportunity to insert `Eve` into their own `k-buckets` if their corresponding bucket is not already full.

Reviewer 2: ```suggestion
To join the DHT new computer, which we will call `Eve`,  must know at least on IP address of a computer which is in the network. Then `Eve` will perform a lookup of herself to get closest known computers which are already aware of `Eve`. After that `Eve` will perform a _bucket refresh_ for all `k-buckets` from the closest known computers to `Eve`. 

This will populate the `Eve's` routing table based on the correctness of the lookup algorithm and the fact that the routing tables of other computers are sufficiently populated. In addition, all computers that are being queried by `Eve` will become aware of `Eve` in the process and thus will have the opportunity to insert `Eve` into their own `k-buckets` if their corresponding bucket is not already full.
```

In Kademlia, peers are sorted as key-value pairs in the DHT. A peerID is stored as a key and the [address](/concepts/addressing/) (or IP address) of a that peer stored as a value.

Kademlia treats peers as leaves in a binary tree. We store hashes of Peer IDs in leaves, which are represented as binary numbers. Notice, that we do not need to store the whole hash to specify the peer. We can store only prefix if prefix is unique.

Leaves on the tree that are not peers represent [content](/concepts/content-routing/) which is stored by some close nodes.

![Kademlia-system](Kademlia-system.png)

The red node in this diagram has the prefix `001`. Traversing this graph from the top to the bottom along the edges, you will notice a blue node with the prefix `11101`, as well as yellow leaves which are other peers.

For any given node, Kademlia divides the whole binary tree into a series of successively lower subtrees. The highest subtree consists of the half of the binary tree not containing the original node. The next subtree consists of the half of the remaining tree not containing the next highest node, and so on.

> You can see here for node `001` there are three levels of subtrees are also circle, consisting of all nodes with prefixes `1` (highest level subtree) , `01`, `000` (lowest level subtree).

Reviewer 2: ```suggestion
You can see here for node `001` there are three levels of subtrees are also circle, consisting of all nodes with prefixes `1` (highest level subtree) , `01`, and `000` (lowest level subtree).
```

_A brief reminder_: Alice's peerID is a key in the DHT, and the value of this key is Alice's IP address.

Bob can locate Alice in this example. Bob knows Alice's peerID and he needs to get her IP address. From the subtree with the prefix `1` Bob knows a peer with prefix `101`. He sends a message to their address, where he asks for Alice's address.

![1-hop](1-hop.png)

Peer `100` doesn't have Alice's key. But they have key of node `110` from their subtree which locates Alice better! And we are doing second hop:

![2-hop](2-hop.png)

Peer `110` also doesn't have Alice's key, but they again locate Alice better and send request to node with prefix `1111`.

![3-hop](3-hop.png)

Peer `1111` stores Alices peer id and can send the key to the Bob! This is how routing works and here is full scheme:

![final-hop](final-hop.png)

Peer `1111` sends response directly to Bob, since each request in this chain had Bob's address.

### Routing table

The place where one node stores info about other nodes is called a routing table. The routing table is a binary tree whose leaves are `k-buckets`.

### k-buckets

Take a look at the initial tree:
![Kademlia-system](Kademlia-system.png)

> See this circled sub-trees? Remember that the Kademlia protocol ensures that every node knows of (and can locate) at least one node in each of its subtrees. If that subtree contains a node. `k-bucket` of a sub-tree is a set of that nodes. Once again: from each subtree we take only those nodes that we know - and we call it `k-bucket`. In out case `k-buckets` for Bob look like this

Reviewer 2: ```suggestion
See these circled sub-trees? Remember that the Kademlia protocol ensures that every node knows of (and can locate) at least one node in each of its subtrees. If that subtree contains a node. `k-bucket` of a sub-tree is a set of that nodes. Once again: from each subtree we take only those nodes that we know - and we call it `k-bucket`. In out case `k-buckets` for Bob look like this
```

![k-buckets](k-buckets.png)

Each `k-bucket` covers some range of the ID space, and together the `k-buckets` cover the entire peer ID space with no overlap.

## Dynamic DHT

> Okay, that was it for static DHT. And the gap from our current state to a fully dynamic DHT is not as far away as it seems. Let us first consider how we can extend this protocol to support computers sporadically leaving.

Reviewer 2: ```suggestion
The gap from our current state to a fully dynamic DHT is not as far away as it seems. Let us first consider how we can extend this protocol to support computers sporadically leaving.
```

Author: Sorry for taking so much time 😅
And thanks!

Reviewer 2: No worries at all, thanks for contributing and taking the time to write this!

### Peers Leaving

Let us assume that we have some known parameter `k` that represents some given number of computers. This number should be a value such that it is extremely unlikely that all of those computers will leave the network in the same hour. The original Kademlia paper that the `k` value should be 20.

Next, our `k-bucket` should, instead of storing just 1 computer, store `k` computers within that `k-bucket’s` range. Thus, with high likelihood, there will always be at least one computer online in each of a computer’s `k-buckets`.

> And, when hopping, instead of asking one computer from the `k-bucket` we will be asking `k` number of peers. Some may not reply, but not all.

Reviewer 3: I believe this isn't all `k` peers but some other number `α` of conucrrent reqs we make

> And, when hopping, instead of asking one computer from the `k-bucket` we will be asking `k` number of peers. Some may not reply, but not all.

Reviewer 3: I believe this isn't all k peers but some other number α of conucrrent reqs we make
